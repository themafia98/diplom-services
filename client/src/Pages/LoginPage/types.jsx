import PropTypes from 'prop-types';
import { udataType } from 'types';
const { object, bool } = PropTypes;

export const loginType = {
  router: object.isRequired,
  history: object.isRequired,
  location: object.isRequired,
  match: object.isRequired,
  authLoad: bool.isRequired,
  udata: udataType.isRequired,
};
