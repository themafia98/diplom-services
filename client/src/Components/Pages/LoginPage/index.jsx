import React from 'react';
import _ from 'lodash';
import PropTypes from 'prop-types';
import { Redirect, NavLink } from 'react-router-dom';
import { Button, Input } from 'antd';
import { connect } from 'react-redux';
import { showGuile, loadUdata } from '../../../Redux/actions/publicActions';
import { addTabAction, setActiveTabAction } from '../../../Redux/actions/routerActions';

import Logo from '../../Logo';
import ModalWindow from '../../ModalWindow';
import modelContext from '../../../Models/context';

class LoginPage extends React.Component {
  state = {
    loading: false,
    loginAuth: null,
    redirect: false,
    errorMessage: null,
  };

  static contextType = modelContext;

  static propTypes = {
    addTab: PropTypes.func.isRequired,
    router: PropTypes.object.isRequired,
  };

  enterLoading = event => {
    const { addTab, onLoadUdata } = this.props;

    const { Request, config = {} } = this.context;

    const { state: { value: login = '' } = {} } = this.login || {};
    const { state: { value: password = '' } = {} } = this.password || {};

    if (login && password) {
      const rest = new Request();
      this.setState({ errorMessage: null, loading: true });
      rest
        .sendRequest(
          '/login',
          'POST',
          {
            email: login,
            password,
          },
          false,
        )
        .then(res => {
          if (res.status === 200 && res.data && res.data['user'].token) {
            localStorage.setItem('token', JSON.stringify(res.data['user'].token));
            this.setState({
              loginAuth: true,
            });
            const udataObj = Object.keys(res.data['user']).reduce((accumulator, key) => {
              if (key !== 'token') {
                accumulator[key] = res.data['user'][key];
              }
              return accumulator || {};
            }, {});

            let path = 'mainModule';
            const defaultModule = config.menu.find(item => item['SIGN'] === 'default');
            if (defaultModule) path = defaultModule.EUID;

            onLoadUdata(udataObj);

            addTab(path);
          } else throw new Error(res.status);
        })
        .catch(error => {
          const {
            response: { data = '' },
          } = error || {};
          debugger;
          const isHtml = /DOCTYPE html/gi.test(data);
          const isValid = _.isString(data) && data && !isHtml;

          const errorMessage = isValid ? data : 'Ошибка авторизации';

          this.setState({
            errorMessage,
            loading: false,
          });
        });
    }
  };

  getCurrentUser = () => {
    return this.state.user;
  };

  login = null;
  password = null;

  refLogin = node => (this.login = node);
  refPassword = node => (this.password = node);

  render() {
    const { refLogin, refPassword, enterLoading } = this;
    const { authLoad = false } = this.props;
    const { loading, errorMessage, loginAuth } = this.state;
    const { config = {} } = this.context;

    if (authLoad || loginAuth) return <Redirect to="/dashboard" />;

    return (
      <div className="loginPage">
        <div className="loginPage__loginContainer">
          <h1 className="loginContainer__title">{config['title']}</h1>
          <Logo />
          <form method="POST" name="loginForm" className="loginContainer__loginForm">
            <div className="notificationWrapper">
              {errorMessage ? <p className="errorMessage">{errorMessage}</p> : null}
            </div>
            <Input name="email" aria-label="login" size="large" placeholder="login" ref={refLogin} />
            <Input
              aria-label="password"
              type="password"
              name="password"
              size="large"
              placeholder="password"
              ref={refPassword}
            />
            <Button
              aria-label="login-button"
              className="enterSystem"
              type="primary"
              loading={loading}
              onClick={enterLoading}
            >
              Войти
            </Button>
            <ModalWindow mode="reg" />
            <NavLink aria-label="recovory-link" className="recovory-link" to="/recovory">
              Восстановление доступа
            </NavLink>
          </form>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => {
  const { udata = {} } = state.publicReducer || {};
  return {
    router: { ...state.router },
    udata,
  };
};

const mapDispatchToProps = dispatch => {
  return {
    addTab: tab => dispatch(addTabAction(tab)),
    onShowGuide: show => dispatch(showGuile(show)),
    setCurrentTab: tab => dispatch(setActiveTabAction(tab)),
    onLoadUdata: async udata => await dispatch(loadUdata(udata)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(LoginPage);
export { LoginPage };
