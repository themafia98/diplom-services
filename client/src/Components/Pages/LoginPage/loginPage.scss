.loginPage {
    display: flex;
    justify-content: center;
    align-items: center;
    width: 100%;
    height: 100vh;
    min-height: 100%;
    max-width: 1140px;
    margin: 0 auto;

    .loginPage__loginContainer {
        display: flex;
        flex-direction: column;
        align-items: center;
        width: 60%;
        height: 70vh;
        box-shadow: 0 0 3px;
        padding: 5px;
        background-color: $snow_l1;

        .loginContainer__title {
            text-align: center;
        }

        .loginContainer__logo {
            display: flex;
            justify-content: center;
            img {
                max-width: 100px;
                width: 100%;
                height: 100%;
            }
        }
        .loginContainer__loginForm {
            display: flex;
            width: 60%;
            flex-direction: column;
            margin-top: 10%;

            input,
            button {
                width: 100%;
                height: 50px;
                text-align: center;
                margin-top: 10px;
                padding: 0;
            }

            button {
                font-weight: bold;
            }
        }
    }
}
