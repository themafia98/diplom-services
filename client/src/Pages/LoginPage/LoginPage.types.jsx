import PropTypes from 'prop-types';
const { object, bool, func } = PropTypes;

export const loginType = {
  location: object,
  authLoad: bool.isRequired,
  initialSession: func.isRequired,
};
