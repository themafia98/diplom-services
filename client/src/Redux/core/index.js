export { default } from './Core.namespace';
