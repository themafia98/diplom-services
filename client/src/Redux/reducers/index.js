import { combineReducers } from 'redux';
import publicReducer from './publicReducer';
import router from './routerStateReducer';
import socketReducer from './socketReducer';
import systemReducer from './systemReducer';

export default combineReducers({
  publicReducer,
  router,
  socketReducer,
  systemReducer,
});
