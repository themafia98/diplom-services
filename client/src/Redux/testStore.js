import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

/** stores for  jest */

export const initialState = {
  tabReducer: {
    parentSize: null,
    childrenSize: null,
    flag: false,
  },
  publicReducer: {
    status: 'online',
    prewStatus: 'online',
    requestError: null,
    caches: {},
  },
  router: {
    currentActionTab: 'mainModule',
    actionTabs: ['MainModule'],
    routeDataActive: {},
    routeData: {},
  },
};

const middlewares = [thunk.withExtraArgument({})]; /** test middlewares */
const mockStore = configureStore(middlewares); /** test config store */

const store = mockStore(initialState); /** init test store */

export default store;
