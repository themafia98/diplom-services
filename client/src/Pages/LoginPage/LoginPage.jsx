import React, { useState, useCallback, useContext, useRef } from 'react';
import { loginType } from './LoginPage.types';
import { Redirect, NavLink } from 'react-router-dom';
import { Button, Input } from 'antd';
import { useSelector } from 'react-redux';
import Logo from 'Components/Logo';
import ModalWindow from 'Components/ModalWindow';
import ModelContext from 'Models/context';
import { SIGNS } from 'App.constant';

const LoginPage = ({ initialSession, authLoad, location }) => {
  const { appConfig } = useSelector((state) => {
    const { appConfig } = state.publicReducer;
    return {
      appConfig,
    };
  });

  const { pathname = '' } = location || {};
  const { regInclude = true, recovoryInclude = true } = appConfig || {};

  const [loading, setLoading] = useState(false);
  const [loginAuth, setAuth] = useState(null);
  const [errorMessage, setErrorMessage] = useState(null);

  const loginRef = useRef(null);
  const passwordRef = useRef(null);

  const context = useContext(ModelContext);

  const enterLoading = useCallback(async () => {
    const { Request } = context;

    const { menu = [] } = appConfig || {};
    const { current: loginNode = null } = loginRef || {};
    const { current: passwordNode = null } = passwordRef || {};

    const { state: { value: loginValue = '' } = {} } = loginNode || {};
    const { state: { value: passwordValue = '' } = {} } = passwordNode || {};

    if (!loginValue || !passwordValue) {
      return;
    }

    const rest = new Request();

    setErrorMessage(null);
    setLoading(true);

    try {
      const response = await rest.sendRequest(
        '/login',
        'POST',
        {
          email: loginValue,
          password: passwordValue,
        },
        false,
      );

      if (response.status !== 200) {
        throw new Error(`${response.status}`);
      }

      const { user: udata, token } = response.data;

      if (!token) {
        throw new Error('Bad user auth token');
      }

      if (!udata) {
        throw new Error('Invalid udata');
      }

      localStorage.setItem('token', token);

      const defaultModule = menu?.find((item) => item?.SIGN === SIGNS.DEFAULT_SIGN);

      initialSession(udata, defaultModule?.EUID || 'mainModule');
      setAuth(true);
    } catch (error) {
      const { data = '' } = error?.response || {};

      const parser = new DOMParser();
      const doc = parser.parseFromString(data, 'text/html');
      const isHtml = Array.from(doc.body.childNodes).some((node) => node.nodeType === 1);
      const isValid = data && typeof data === 'string' && !isHtml;

      const errorMessage = isValid
        ? data
        : isHtml
        ? 'Ошибка сервера, попробуйте позже.'
        : 'Ошибка авторизации';

      setErrorMessage(errorMessage);
      setLoading(false);
    }
  }, [appConfig, context, initialSession]);

  const onKeyDown = ({ key = '' }) => {
    if (key === 'Enter') enterLoading();
  };

  if (authLoad || loginAuth) {
    return <Redirect to="/dashboard" />;
  }

  if ((!authLoad || !loginAuth) && pathname !== '/') {
    return <Redirect to="/" />;
  }

  return (
    <div className="loginPage">
      <div className="loginPage__loginContainer">
        <h1 className="loginContainer__title">{appConfig?.title}</h1>
        <Logo />
        <form onKeyDown={onKeyDown} method="POST" name="loginForm" className="loginContainer__loginForm">
          <div className="notificationWrapper">
            {errorMessage ? <p className="errorMessage">{errorMessage}</p> : null}
          </div>
          <Input name="email" aria-label="login" size="large" placeholder="login" ref={loginRef} />
          <Input
            aria-label="password"
            type="password"
            name="password"
            size="large"
            autoComplete=""
            placeholder="password"
            ref={passwordRef}
          />
          <Button
            aria-label="login-button"
            className="enterSystem"
            type="primary"
            loading={loading}
            onClick={enterLoading}
          >
            Войти
          </Button>
          {regInclude ? <ModalWindow mode="reg" /> : null}
          {recovoryInclude ? (
            <NavLink aria-label="recovory-link" className="recovory-link" to="/recovory">
              Восстановление доступа
            </NavLink>
          ) : null}
        </form>
      </div>
    </div>
  );
};

LoginPage.propTypes = loginType;
LoginPage.defaultProps = {
  location: {},
};

export default LoginPage;
