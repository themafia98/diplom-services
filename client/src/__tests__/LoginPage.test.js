import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';

import { LoginPage } from '../Components/Pages/LoginPage/LoginPage';
import { initialState } from '../Redux/testStore';

describe('<LoginPage /> template', () => {
  test('Should be render', () => {
    const props = {
      addTab: () => {},
      router: {},
      ...initialState,
    };

    const LoginPageWrapper = shallow(<LoginPage {...props} />);
    expect(toJson(LoginPageWrapper)).toMatchSnapshot();

    expect(LoginPageWrapper.find('.enterSystem').simulate('click', [])).toBeTruthy();
    expect(toJson(LoginPageWrapper)).toMatchSnapshot();

    LoginPageWrapper.setState({
      loading: true,
    });

    expect(LoginPageWrapper.find('.enterSystem').prop('loading')).toEqual(true);
    expect(toJson(LoginPageWrapper)).toMatchSnapshot();

    expect(LoginPageWrapper.find('NavLink').simulate('click', [])).toBeTruthy();
    expect(toJson(LoginPageWrapper)).toMatchSnapshot();
  });
});
