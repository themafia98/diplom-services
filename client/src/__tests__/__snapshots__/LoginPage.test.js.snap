// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginPage /> template Should be render 1`] = `
<div
  className="loginPage"
>
  <div
    className="loginPage__loginContainer"
  >
    <h1
      className="loginContainer__title"
    />
    <Logo />
    <form
      className="loginContainer__loginForm"
      method="POST"
      name="loginForm"
    >
      <div
        className="notificationWrapper"
      />
      <Input
        aria-label="login"
        name="email"
        placeholder="login"
        size="large"
        type="text"
      />
      <Input
        aria-label="password"
        name="password"
        placeholder="password"
        size="large"
        type="password"
      />
      <Button
        aria-label="login-button"
        block={false}
        className="enterSystem"
        ghost={false}
        htmlType="button"
        loading={false}
        onClick={[Function]}
        type="primary"
      >
        Войти
      </Button>
      <ModalWindow
        mode="reg"
      />
      <NavLink
        aria-label="recovory-link"
        className="recovory-link"
        to="/recovory"
      >
        Восстановление доступа
      </NavLink>
    </form>
  </div>
</div>
`;

exports[`<LoginPage /> template Should be render 2`] = `
<div
  className="loginPage"
>
  <div
    className="loginPage__loginContainer"
  >
    <h1
      className="loginContainer__title"
    />
    <Logo />
    <form
      className="loginContainer__loginForm"
      method="POST"
      name="loginForm"
    >
      <div
        className="notificationWrapper"
      />
      <Input
        aria-label="login"
        name="email"
        placeholder="login"
        size="large"
        type="text"
      />
      <Input
        aria-label="password"
        name="password"
        placeholder="password"
        size="large"
        type="password"
      />
      <Button
        aria-label="login-button"
        block={false}
        className="enterSystem"
        ghost={false}
        htmlType="button"
        loading={false}
        onClick={[Function]}
        type="primary"
      >
        Войти
      </Button>
      <ModalWindow
        mode="reg"
      />
      <NavLink
        aria-label="recovory-link"
        className="recovory-link"
        to="/recovory"
      >
        Восстановление доступа
      </NavLink>
    </form>
  </div>
</div>
`;

exports[`<LoginPage /> template Should be render 3`] = `
<div
  className="loginPage"
>
  <div
    className="loginPage__loginContainer"
  >
    <h1
      className="loginContainer__title"
    />
    <Logo />
    <form
      className="loginContainer__loginForm"
      method="POST"
      name="loginForm"
    >
      <div
        className="notificationWrapper"
      />
      <Input
        aria-label="login"
        name="email"
        placeholder="login"
        size="large"
        type="text"
      />
      <Input
        aria-label="password"
        name="password"
        placeholder="password"
        size="large"
        type="password"
      />
      <Button
        aria-label="login-button"
        block={false}
        className="enterSystem"
        ghost={false}
        htmlType="button"
        loading={true}
        onClick={[Function]}
        type="primary"
      >
        Войти
      </Button>
      <ModalWindow
        mode="reg"
      />
      <NavLink
        aria-label="recovory-link"
        className="recovory-link"
        to="/recovory"
      >
        Восстановление доступа
      </NavLink>
    </form>
  </div>
</div>
`;

exports[`<LoginPage /> template Should be render 4`] = `
<div
  className="loginPage"
>
  <div
    className="loginPage__loginContainer"
  >
    <h1
      className="loginContainer__title"
    />
    <Logo />
    <form
      className="loginContainer__loginForm"
      method="POST"
      name="loginForm"
    >
      <div
        className="notificationWrapper"
      />
      <Input
        aria-label="login"
        name="email"
        placeholder="login"
        size="large"
        type="text"
      />
      <Input
        aria-label="password"
        name="password"
        placeholder="password"
        size="large"
        type="password"
      />
      <Button
        aria-label="login-button"
        block={false}
        className="enterSystem"
        ghost={false}
        htmlType="button"
        loading={true}
        onClick={[Function]}
        type="primary"
      >
        Войти
      </Button>
      <ModalWindow
        mode="reg"
      />
      <NavLink
        aria-label="recovory-link"
        className="recovory-link"
        to="/recovory"
      >
        Восстановление доступа
      </NavLink>
    </form>
  </div>
</div>
`;
